Crashlytics/Crashlytics/Controllers/FIRCLSExistingReportManager.m

```diff
@@ -45,51 +52,106 @@ - (instancetype)initWithManagerData:(FIRCLSManagerData *)managerData
+  return self.existingUnemptyActiveReportPaths.count;
 }
 
-- (int)countSubmittableAndDeleteUnsubmittableReportPaths:(NSArray *)reportPaths {
-  int count = 0;
+- (NSArray *)getUnemptyExistingActiveReportsAndDeleteEmpty:(NSArray *)reportPaths {
+  NSMutableArray *unemptyReports = [NSMutableArray array];
   for (NSString *path in reportPaths) {
     FIRCLSInternalReport *report = [FIRCLSInternalReport reportWithPath:path];
-    if ([report needsToBeSubmitted]) {
-      count++;
+    if ([report hasAnyEvents]) {
+      [unemptyReports addObject:path];
     } else {
       [self.operationQueue addOperationWithBlock:^{
-        [self->_fileManager removeItemAtPath:path];
+        [self.fileManager removeItemAtPath:path];
       }];
     }
   }
-  return count;
+  return unemptyReports;
 }
 
-- (void)processExistingReportPaths:(NSArray *)reportPaths
-               dataCollectionToken:(FIRCLSDataCollectionToken *)dataCollectionToken
+- (void)sendUnsentReportsWithToken:(FIRCLSDataCollectionToken *)dataCollectionToken
                           asUrgent:(BOOL)urgent {
-  for (NSString *path in reportPaths) {
+  for (NSString *path in self.existingUnemptyActiveReportPaths) {
     [self processExistingActiveReportPath:path
                       dataCollectionToken:dataCollectionToken
                                  asUrgent:urgent];
   }
+
+  // deal with stuff in processing more carefully - do not process again
+  [self.operationQueue addOperationWithBlock:^{
+    for (NSString *path in self.processingReportPaths) {
+      FIRCLSInternalReport *report = [FIRCLSInternalReport reportWithPath:path];
+      [self.reportUploader prepareAndSubmitReport:report
+                              dataCollectionToken:dataCollectionToken
+                                         asUrgent:NO
+                                   withProcessing:NO];
+    }
+  }];
+
+  // Because this could happen quite a bit after the inital set of files was
+  // captured, some could be completed (deleted). So, just double-check to make sure
+  // the file still exists.
+  [self.operationQueue addOperationWithBlock:^{
+    for (NSString *path in self.preparedReportPaths) {
+      if (![[self.fileManager underlyingFileManager] fileExistsAtPath:path]) {
+        continue;
+      }
+      [self.reportUploader uploadPackagedReportAtPath:path
+                                  dataCollectionToken:dataCollectionToken
+                                             asUrgent:NO];
+    }
+  }];
 }
 
 - (void)processExistingActiveReportPath:(NSString *)path
                     dataCollectionToken:(FIRCLSDataCollectionToken *)dataCollectionToken
                                asUrgent:(BOOL)urgent {
   FIRCLSInternalReport *report = [FIRCLSInternalReport reportWithPath:path];
 
-  // TODO: needsToBeSubmitted should really be called on the background queue.
-  if (![report needsToBeSubmitted]) {
+  // TODO: hasAnyEvents should really be called on the background queue.
```

Comment:
What was the reasoning for this not being on a background queue?

Reply:
We humaned on this - might move to a background thread if possible